### PR TITLE
3943 improve column autosizing

### PR DESCRIPTION
**Pull request by utas-raymondng**

### CODE REVIEW

File: web/src/components/data-entities/diver/DiverList.js [hunk 1]
```diff
@@ -1,103 +1,129 @@
...
 
-  useEffect(() => {
-    async function fetchDivers() {
-      await getResult('divers').then((res) => setRowData(res.data));
-    }
-    fetchDivers();
-  }, []);
+    // Auto size function to be call each time data changed, so the grid always autofit
+    const autoSizeAll = (skipHeader) => {
+        if (dGridRef.current != null) {
```
Comment (nspool) [inline web/src/components/data-entities/diver/DiverList.js:19]: should be `typeof dGridRef !== "null"`, but you can also use null coalescing to write at dGridRef?.current, or just use a truthy check `if(dGridRef.current) {`

File: web/src/components/data-entities/diver/DiverList.js [hunk 2]
```diff
@@ -1,103 +1,129 @@
...
 import 'ag-grid-enterprise';
 
 const DiverList = () => {
-  const [gridApi, setGridApi] = useState();
-  const [delta, setDelta] = useState([]);
-  const [rowData, setRowData] = useState();
-  const [errors, setErrors] = useState([]);
+    const [delta, setDelta] = useState([]);
+    const [errors, setErrors] = useState([]);
+    const dGridRef = useRef(null);
```
Comment (nspool) [inline web/src/components/data-entities/diver/DiverList.js:15]: What is the naming convention for dGridRef (also oGridRef and lGridRef) ?

File: web/src/components/data-entities/diver/DiverList.js [hunk 3]
```diff
@@ -1,103 +1,129 @@
...
-    if (gridApi) gridApi.redrawRows();
-  }, [gridApi, delta, errors]);
+    const onGridReady = (event) => {
+        async function fetchDivers(event) {
+            await getResult(constants.diverList.URL).then(
+                (res) => {
+                    // Use setRowData in api will not trigger onGridReady but onDataChange event.
+                    // if you use useState and connect row to setRowData then you will
+                    // keep fire onGridReady as row change
+                    console.debug('Loading diver dataset..');
```
Comment (nspool) [inline web/src/components/data-entities/diver/DiverList.js:31]: Please remove `console.debug` statements.

File: web/src/components/data-entities/diver/DiverList.js [hunk 4]
```diff
@@ -1,103 +1,129 @@
...
+            autoSizeAll(false);
+        });
+    };
 
-  const chooseCellStyle = (params) => {
-    if (params.context.errors.some((e) => e.id === params.data.diverId)) return {backgroundColor: red[100]};
-    if (params.context.delta[params.data.diverId]) return {backgroundColor: grey[100]};
-    return null;
-  };
+    // Why need this extra redraw?
```
Comment (nspool) [inline web/src/components/data-entities/diver/DiverList.js:46]: Just delete code that isn't necessary. Avoid leaving code commented out if possible.

File: web/src/common/constants.js [hunk 5]
```diff
@@ -214,3 +214,10 @@ export const extendedMeasurements = Object.freeze([
 ]);
 
 export const allMeasurements = [...measurements, ...extendedMeasurements];
+
+export const constants = Object.freeze({
```
Comment (nspool) [inline web/src/common/constants.js:218]: Avoid adding constants that are only used once. Constants for single words like `divers` probably isn't useful.

File: web/src/components/data-entities/diver/DiverList.js [hunk 6]
```diff
@@ -1,103 +1,129 @@
...
 
-  const onCellValueChanged = (e) => {
-    setDelta((data) => {
-      const newData = {...data};
-      newData[e.data.diverId] = e.data;
-      return newData;
-    });
-  };
+        fetchDivers(event).then(() => {
+            // Now we have the data to do auto sizing, however the build in function only auto size visible rows,
```
Comment (nspool) [inline web/src/components/data-entities/diver/DiverList.js:39]: No need to comment on what the code is doing unless it isn't obvious (or needs some extra explaining).

File: web/src/components/ui/AlertDialog.test.tsx [hunk 7]
```diff
@@ -1,14 +1,15 @@
-import React from 'react'
```
Comment (nspool) [inline web/src/components/ui/AlertDialog.test.tsx:1]: Try to avoid changing unrelated files.

File: web/src/components/data-entities/diver/__test__/DriverList.test.tsx [hunk 8]
```diff
@@ -0,0 +1,63 @@
+// @ts-ignore
```
Comment (nspool) [inline web/src/components/data-entities/diver/__test__/DriverList.test.tsx:1]: Should be `DiverList.test.tsx`. Other tests are kept in the same directory as the component instead of in `__tests__`.

File: web/src/components/data-entities/diver/__test__/DriverList.test.tsx [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        mockGetResult = jest.spyOn(axiosInstance, 'getResult');
+    });
+
+    afterEach(() => {
+        mockGetResult.mockRestore();
+    });
+
+    it('grid columns auto sizing after data load', async () => {
+        // Load sample data.
+        const canned = require('./diver.json');
```
Comment (nspool) [inline web/src/components/data-entities/diver/__test__/DriverList.test.tsx:27]: Please remove this file and use sample data. Don't use real data for tests since this is a public repo.

**Review by nspool — CHANGES_REQUESTED** [review]
Also, no need to add issue number or filenames in the first line of the git commit. Anything extra can go in the body.

File: web/src/components/data-entities/diver/DiverList.js (hunk 2 above)
Comment (utas-raymondng) [reply web/src/components/data-entities/diver/DiverList.js:15]: Change to have name alighted

File: web/src/components/data-entities/diver/DiverList.js (hunk 1 above)
Comment (utas-raymondng) [reply web/src/components/data-entities/diver/DiverList.js:19]: Updated

File: web/src/components/data-entities/diver/DiverList.js (hunk 6 above)
Comment (utas-raymondng) [reply web/src/components/data-entities/diver/DiverList.js:39]: Removed

File: web/src/components/ui/AlertDialog.test.tsx (hunk 7 above)
Comment (utas-raymondng) [reply web/src/components/ui/AlertDialog.test.tsx:1]: I got a couple of alert in Intellj complaining about the use of " and '

File: web/src/components/data-entities/diver/__test__/DriverList.test.tsx (hunk 8 above)
Comment (utas-raymondng) [reply web/src/components/data-entities/diver/__test__/DriverList.test.tsx:1]: Fixed typo

File: web/src/components/data-entities/diver/__test__/DriverList.test.tsx (hunk 9 above)
Comment (utas-raymondng) [reply web/src/components/data-entities/diver/__test__/DriverList.test.tsx:27]: Done

File: web/src/common/constants.js (hunk 5 above)
Comment (utas-raymondng) [reply web/src/common/constants.js:218]: I can understand why you think the constants is use in one place only because I forgot a check in the test case that it should return data only if the url matches, I added back the if check in the test

File: web/src/components/data-entities/diver/DiverList.js (hunk 3 above)
Comment (utas-raymondng) [reply web/src/components/data-entities/diver/DiverList.js]: Removed

File: web/src/components/data-entities/diver/DiverList.js (hunk 4 above)
Comment (utas-raymondng) [reply web/src/components/data-entities/diver/DiverList.js]: Done

File: web/src/components/data-entities/diver/__test__/DiverList.test.tsx [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+            .then(() => {
+                // For debug to tell exactly what is inside.
+                logRoles(screen.getByRole('grid'));
+
+                const gridcells = screen.getAllByRole('gridcell');
+
+                // Without adjust the size of browser the number of row display is 12, since we have two column so it is 24
+                expect(gridcells).toHaveLength(24);
+                expect(screen.getByText('TEST11')).toBeInTheDocument();   // Last diver on the screen
+                expect(screen.getByText('TEST11')).toHaveStyle({width: '48px'});   // Last diver on the screen
```
Comment (nspool) [inline web/src/components/data-entities/diver/__test__/DiverList.test.tsx:63]: Please just remove this test for now, since we can't really test the sizing.

**Review by nspool — CHANGES_REQUESTED** [review]
Please also commit `yarn.lock`

File: web/src/components/data-entities/diver/__test__/DiverList.test.tsx (hunk 10 above)
Comment (utas-raymondng) [reply web/src/components/data-entities/diver/__test__/DiverList.test.tsx]: Removed

File: web/src/common/constants.js (hunk 5 above)
Comment (utas-raymondng) [reply web/src/common/constants.js]: Removed given the test case deleted

File: web/src/components/data-entities/diver/__test__/DriverList.test.tsx (hunk 8 above)
Comment (utas-raymondng) [reply web/src/components/data-entities/diver/__test__/DriverList.test.tsx]: Removed test

**Comment by utas-raymondng** [comment]
Diver page auto sizing removed

File: web/src/components/data-entities/site/SiteList.js [hunk 11]
```diff
@@ -10,152 +10,173 @@ import {entityDelete} from '../../../api/api';
 import 'ag-grid-enterprise';
 
 const SiteList = () => {
-  const [gridApi, setGridApi] = useState();
-  const [redirect, setRedirect] = useState();
-  const [dialogState, setDialogState] = useState({open: false});
+    const [redirect, setRedirect] = useState();
```
Comment (nspool) [inline web/src/components/data-entities/site/SiteList.js:13]: Tab size should be set to 2. Looks like it's set to 4?

**Review by nspool — CHANGES_REQUESTED** [review]
Looks ready to merge. But can you please first reformat the JS files so that they have a tab spacing of 2 instead of 4, and also remove the "axios-mock-adapter" and "@jest/globals" devDependencies?

File: web/src/components/data-entities/site/SiteList.js (hunk 11 above)
Comment (utas-raymondng) [reply web/src/components/data-entities/site/SiteList.js]: Updated and removed dependencies

File: web/src/components/data-entities/observable-item/ObservableItemList.js [hunk 12]
```diff
@@ -79,17 +94,17 @@ const ObservableItemList = () => {
...
-        <AgGridColumn width={100} field="supersededBy" />
-        <AgGridColumn width={100} field="supersededNames" />
-        <AgGridColumn width={100} field="supersededIDs" />
-        <AgGridColumn width={100} field="phylum" />
-        <AgGridColumn width={100} field="class" />
-        <AgGridColumn width={100} field="order" />
-        <AgGridColumn width={100} field="family" />
-        <AgGridColumn width={100} field="genus" />
+        <AgGridColumn field="typeName" headerName="Type" />
+        <AgGridColumn cellStyle={{'min-width': '200px'}} field="name" />
```
Comment (nspool) [inline web/src/components/data-entities/observable-item/ObservableItemList.js:98]: AgGridColumn has a prop called `minWidth`. No need to set use CSS.

**Review by nspool — APPROVED** [review]
